### ISSUE TITLE
[HOLD for payment 2025-01-02] [$500] [Simple AA in NewDot] Update the approval chain / next receiver logic to handle correct optimistic next approver in Advanced Approval cases

### ISSUE THREAD

**Issue by Beamanator** [Expensify/App#52458]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** Current
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Expensify/Expensify Issue URL:** https://github.com/Expensify/Expensify/issues/430318

## [1] Action Performed:

1. Create a workspace in NewDot
2. Enable delay submission
3. Invite submitter, Approver A, Approver B
4. Set to Advanced Approval in OldDot via workspace owner (workspace should be type Control)
5. Set category approvals for Category A going to Approver A + Category B as Approver B
6. As submitter in NewDot, submit 2 expenses to the WS chat, one with Category A, one with Category B
7. Submit the report
8. Confirm first approver is Approver A
9. As Approver A, approve the report

## [1] Expected Result:

10. Confirm the second approver is Approver B (**both offline & online**)
11. As Approver B, approve the report
12. Confirm the third approver is workspace owner (**both offline & online**)
13. As WS owner, approve and confirm that the report is final approved.

## [2] Action Performed:

1. Create a workspace in NewDot
2. Enable delay submission
3. Invite submitter, Approver A, Approver B, Approver C
4. Set to Advanced Approval in OldDot via workspace owner (workspace should be type Control)
5. Set up approver workflow like this:
    - Submitter `submitsTo` Approver A
    - Approver A `forwardsTo` Approver B
    - Approver C `forwardsTo` workspace owner
6. As submitter, create an expense in the workspace chat & submit it -> this should get submitted to Approver A
7. As workspace owner, update approver workflow so that Submitter has `submitsTo` of Approver C
8. As Approver A, approve the report (in NewDot)

## [2] Expected Result:

9. Verify that Approver C is the next approver (**both offline & online**)
10. As Approver C, approve the report
11. Verify the next approver is the workspace owner (**both offline & online**)
12. As the workspace owner, approve and confirm the report is final approved

### ReportUtils.getApprovalChain

This function should ONLY return the submitter's approval chain based on the current `policy.employeeList` data.

We need to start including category & tag approvers in this function. Category approvers should come first, then tag approvers, then approvers in `employeeList`.

Here's what we have in C++ which should give a good starting point:

```c++
    vector<string> approvalChain;
    string nextApprover = Policy::getSubmitsToEmail(db, policyID, submitterEmail, employeeList);

    // If there's no submitsTo, just use the policy owner so we have someone to submit to
    if (nextApprover.empty()) {
        const string& policyOwner = Account::getEmail(db, Policy::getOwner(db, policyID));
        approvalChain.push_back(policyOwner);
    }

    // We don't want to add someone who's already in the approval chain, or it'd become an infinite loop
    const int64_t reportTotal = abs(getTransactionTotal(db, reportID, getCurrency(db, reportID)));
    while (!nextApprover.empty() && find(approvalChain.begin(), approvalChain.end(), nextApprover) == approvalChain.end()) {
        approvalChain.push_back(nextApprover);

        // Get the forwardsTo or overLimitForwardsTo (if over the approval limit)
        nextApprover = Policy::getForwardsTo(employeeList, policyID, nextApprover, reportTotal);
    }

    // The full approval chain with rule approvers before submitsTo/forwardsTo approvers and no repeats
    // We use an unordered_set to ensure unique elements while inserting and preserving the order of the original vectors
    vector<string> fullApprovalChain;
    unordered_set<string> uniqueApprovers;
    for (const auto& approver : ruleApprovers) {
        // Don't force the submitter to approve as a rule approver, because it could get confusing for the user
        if (ruleApprovers.contains(submitterEmail)) {
            SINFO("Submitter is also rule approver, skipping adding to approval chain");
            continue;
        }

        // If insertion was successful (i.e., not already in the set), then add to the full approval chain
        if (uniqueApprovers.insert(approver).second) {
            fullApprovalChain.push_back(approver);
        }
    }
    for (const auto& approver : approvalChain) {
        if (uniqueApprovers.insert(approver).second) {
            fullApprovalChain.push_back(approver);
        }
    }

    // If the submitter of the report is also the last person that needs to approve and the policy doesn't allow self-approval, then skip that person.
    if (!fullApprovalChain.empty() && fullApprovalChain.back() == submitterEmail && Policy::isPreventSelfApprovalEnabled(db, policyID)) {
        fullApprovalChain.pop_back();
    }

    return fullApprovalChain;
```

### IOU.getNextApproverAccountID

This function should figure out who, in the policy's current approvalChain (built by buildApproverChain above) is next in line to approve the report.

This function will also check IF the report has been previously approved by someone in the approval chain - if so, they will be skipped.

Here's what we have in C++ which should give a good starting point:

```c++
string Report::getNextReceiver(SQLite& db, const int64_t reportID, const string& currentApprover, const vector<string>& approvalChain)
{
    if (!verifyState(db, reportID, STATE_OPEN) && !verifyState(db, reportID, STATE_SUBMITTED)) {
        return "";
    }

    const set<int64_t> previousReportApprovers = getAccountsThatHaveApprovedReport(db, reportID);

    // Loop through approval chain to find the next approver who has not yet approved the report
    const string& managerOnVacation = getValue(db, reportID, NVP_MANAGER_ON_VACATION);
    for (auto it = approvalChain.begin(); it != approvalChain.end(); ++it) {
        const string& approver = *it;
        string approverOrDelegate = approver;

        // Handle vacation delegate
        if (approver == managerOnVacation) {
            const string& vacationDelegate = Account::getVacationDelegate(db, managerOnVacation);
            approverOrDelegate = !vacationDelegate.empty() ? vacationDelegate : approver;
        }

        // Return early if the report is open
        if (verifyState(db, reportID, STATE_OPEN)) {
            return approverOrDelegate;
        }

        // Return the first approver in the approval chain that hasn't approved yet
        const bool reportHasBeenApprovedByApproverOrDelegate = previousReportApprovers.contains(Account::getID(db, approverOrDelegate));
        if (currentApprover != approverOrDelegate && !reportHasBeenApprovedByApproverOrDelegate) {
            return approverOrDelegate;
        }
    }

    return "";
}

set<int64_t> Report::getAccountsThatHaveApprovedReport(SQLite& db, const int64_t reportID)
{
    const set<string> RESET_APPROVAL_ACTIONS = {
        ACTION_REJECTED_TO_SUBMITTER,
        ACTION_RETRACTED,
        ReportAction::ACTION_SUBMITTED,
        ReportAction::ACTION_DELEGATE_SUBMIT,
        ACTION_REOPENED,
        ReportAction::ACTION_UNAPPROVED,
    };

    const set<string> APPROVAL_ACTIONS = {
        ReportAction::ACTION_APPROVED,
        ACTION_FORWARDED,
        ACTION_CLOSED,
    };

    set<string> allActionsOfInterest;
    allActionsOfInterest.insert(RESET_APPROVAL_ACTIONS.begin(), RESET_APPROVAL_ACTIONS.end());
    allActionsOfInterest.insert(APPROVAL_ACTIONS.begin(), APPROVAL_ACTIONS.end());
    allActionsOfInterest.insert(ACTION_REJECTED);

    SQResult result;
    DB::read(db, "SELECT action, accountID, created FROM reportActions WHERE reportID = " + SQ(reportID) + " AND action IN (" + SQList(allActionsOfInterest) + ") ORDER BY created DESC;", result);

    set<int64_t> recentApprovers;
    map<int64_t, int64_t> approversWithApprovalCount;
    for (const auto& row : result.rows) {
        const int64_t accountID = SToInt64(row["accountID"]);
        if (!SContains(approversWithApprovalCount, accountID)) {
            // Note: We don't care about approval order because we only care about the order of the "next approvers"
            // in the approval chain to figure out the next receiver, not the order in which previous approvers approved the report
            recentApprovers.insert(accountID);

            // Initialize this approver in the map
            approversWithApprovalCount[accountID] = 0;
        }
        if (APPROVAL_ACTIONS.contains(row["action"])) {
            approversWithApprovalCount[accountID] += 1;
        } else if (row["action"] == ACTION_REJECTED) {
            approversWithApprovalCount[accountID] -= 1;
        } else {
            // We reached an action that resets the "approval" state of a report, ignore rest
            break;
        }
    }

    // Take out all the approvers who had a net 0 amount of approvals (ex: they approved but then unapproved)
    for (const auto& [accountID, count] : approversWithApprovalCount) {
        if (count <= 0) {
            recentApprovers.erase(accountID);
        }
    }
    return recentApprovers;
}
```

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021856641526591668663</li>
        <li>Upwork Job ID: 1856641526591668663</li>
        <li>Last Price Increase: 2024-12-30</li>
<li>Automatic offers: </li>
<ul>
<li>nkdengineer | Contributor | 104876723</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @puneetlath</details>

**Comment by Beamanator** [comment]
cc @nkdengineer in case you're interested to work on this, since you worked on https://github.com/Expensify/App/issues/51001 recently

**Comment by nkdengineer** [comment]
@Beamanator I'd like to work on this.

**Comment by Beamanator** [comment]
Thank you @nkdengineer 👍 

Please feel free to ask questions here before starting your PR IF you have any, either way I will try to do a very detailed review 🙏

**Comment by nkdengineer** [comment]
Sure.

**Comment by nkdengineer** [comment]
### Update: 

- I'm looking into this with the provided logic, I'll finish and raise the PR tomorrow or Saturday.

**Comment by nkdengineer** [comment]
@Beamanator I created a draft [here](https://github.com/Expensify/App/pull/52537). Please help test it with your local backend. Now I can only test with two rule approvers because it's marked as approved after the first rule approver approves the expense report.

**Comment by Beamanator** [comment]
Some problems I've seen so far in my testing
- Note: I have the following setup:
    1. Control policy, on Submit & Approve, default submitter is policy owner
    2. 2 tag approvers & 2 category approver set up
    3. Submitter creates expenses in this order:
        1. Tag approver 1
        2. Category approver 1
        3. Category approver 2
- When Submitter first submits, I see that we submit to Category approver 2, though in it should be Category approver 1 as the first expense w/ a category approver, no? It shouldn't be "the latest category approver", it should be the first by created date.

Can you please try that and see if you reproduce the issue?

As for the approval order when there's multiple category approvers, i think i see the problem in the back end that i'll start working on a fix for

**Comment by nkdengineer** [comment]
> It shouldn't be "the latest category approver", it should be the first by created date.

@Beamanator  I tested this case and `Category approver 1` is the first next approver

**Comment by Beamanator** [comment]
FYI backend PR was merged earlier today, hasn't been deployed yet

**Comment by Beamanator** [comment]
@nkdengineer backend PR was deployed 🙏  please continue with this one & let me know if you hit any more bugs

**Comment by nkdengineer** [comment]
Sure I will retest soon.

**Comment by nkdengineer** [comment]
@Beamanator I tested with the staging server and we still don't archive the expected:

Precondition: 
- A WS with a submitter, approver 1, approver 2, approver 3, approver4
- Enable rules and add a category's approver for approver 1 (category 1), another one for approver 2 (category 2)
- Add a tag's approver for approver 3 (tag 1), another one for approver 4 (tag 2)

1. Create 4 expenses with category 1, category 2, tag 1, tag 2
2. Submit the report
3. After the submit the first approver is approver 1 (`Correct`)
4. As approver 1 approves the report
5. Offline, the next approver is the approver 2 (`Correct`)
6. Online, the next step is waiting admin to pay the expense (`Incorrect`) (The expected should be the next approver is approver 2).

**Comment by Beamanator** [comment]
wtf... i see what you're seeing 🙃 @nkdengineer what approval mode have you been testing on? Me, Submit & Approve



https://github.com/user-attachments/assets/b4dfefb7-9fa5-4c5a-baf6-b95f44963f9d

**Comment by nkdengineer** [comment]
@Beamanator Advance approve mode.

**Comment by Beamanator** [comment]
Sounds good - you should be good @ntdiary as long as you'll be free early next week!

I am working on fixing another internal bug 😬 https://github.com/Expensify/Expensify/issues/446644

**Comment by Beamanator** [comment]
@nkdengineer FYI the bug I found is actually in Submit & Approve flows... Because we DO actually need to test that this all works in both Submit & Approve (S&A) and Advanced Approval (AA) flows 😬 

So once again, hopefully my fix will get merged today, not sureeeee if it'll get deployed by the weekend or not, but if not we'll have to pick this back up Monday 🙃 

In the meantime, I would appreciate if you can let me know your exact setup for the Advanced Approval flow where you found a bug - cuz i don't thinkkk my fix will cover that case, only Submit & Approve

**Comment by Beamanator** [comment]
Okkkk that makes much more sense - thanks @nkdengineer 🙏 

can you please test again on Advance Approval, and if it's working we can try to get the PR ready for that case? Hopefully today / tomorrow I'll get the fix in for `BASIC` approvals

**Comment by nkdengineer** [comment]
Will re-test today.

**Comment by Beamanator** [comment]
Amazing, that's great to hear 👍 👍

**Comment by Beamanator** [comment]
FYI backend PR merged now, may not get deployed till tomorrow - we'll see

**Comment by Beamanator** [comment]
FYI backend PR merged now, may not get deployed till tomorrow - we'll see

**Comment by nkdengineer** [comment]
@ntdiary The [PR](https://github.com/Expensify/App/pull/52537) is ready for review. cc @Beamanator

**Comment by Beamanator** [comment]
Ooohhh goood question - I see that we should return updated Next Steps for the approver, but we may not live update that for all clients when we forward / approve 🤔  Thanks for reporting that! I will make sure to look into that

**Comment by Beamanator** [comment]
Ok I just discussed internally & we found the expected category vs tag approver sort order for NewDot [here](https://expensify.slack.com/archives/C06ML6X0W9L/p1732726934547229?thread_ts=1732725856.867159&cid=C06ML6X0W9L), and here's what we came up with:

Category vs Tag approver order (for now):
1. Category approvers come before tag approvers
2. If there's multiple category approvers for the report (for example, but same logic for tag approvers)
    1. First choose approvers from expenses with earliest (farthest in the past) transaction created date (the selected date on the expense)
    2. If there's multiple expenses on that day w/ different category approvers, sort by earliest transaction inserted date
3. If there's multiple tag approvers for any 1 single expense...
    1. Sort approvers alphabetically

**Comment by nkdengineer** [comment]
@Beamanator Currently, the transaction data in the frontend only have the `created` (the selected date on the expense), we don't have `inserted` date. It makes it a bit hard to sort the transaction list if multiple transactions have the same `created`. Because in this case, we need to compare the iou created action.

**Comment by Beamanator** [comment]
Ooh good shout @nkdengineer - I will fix that in a PR today which should get merged tomorrow

**Comment by nkdengineer** [comment]
Good, I will complete the order after the PR is deployed.

**Comment by Beamanator** [comment]
I was OOO yesterday :D but the PR was deployed 👍

**Comment by nkdengineer** [comment]
Updated the PR.

**Comment by Beamanator** [comment]
welcome!! 🤣

**Comment by Beamanator** [comment]
@puneetlath I'll need you for payment once this is done? Or I can find someone else 🤷

**Comment by puneetlath** [comment]
Ahh I see. Thought this was back-end. Re-assigned.

**Comment by Beamanator** [comment]
Nooooooo worries sir, I'm commenting a lot but yeah it's front end 👍 🙏

**Comment by nkdengineer** [comment]
> I definitely think it's fair to bump the bounty on this one, it's a pretty complicated issue that we're trying to get perfect 👍

@Beamanator Based on [this](https://github.com/Expensify/App/pull/52537#issuecomment-2515708877), please help update the price. Thanks. cc @puneetlath

**Comment by Beamanator** [comment]
Thanks for bringing that up - @puneetlath I agreed that this PR is pretty complex & handles a LOT of approval use cases (both category & tag approvers, plus simple & advanced approval flows), so I believe both @nkdengineer and @ntdiary deserve bumps for this one.

**Comment by puneetlath** [comment]
What should the payout be?

**Comment by Beamanator** [comment]
I would think doubling to $500 is fair

**Comment by puneetlath** [comment]
Ok sounds good. @nkdengineer will be paid $500 here. And @ntdiary will be paid via NewDot. 

@ntdiary can you please complete the checklist?

**Comment by puneetlath** [comment]
@Beamanator what do you think?

**Comment by Beamanator** [comment]
I def like the idea of automated unit tests 👍👍👍👍👍

Though I'm not quite sure how the current process works for that - should I make a new issue & assign one of you to write all the tests? I don't think we would need both a contributor & a C+ for that right? 🤷 

But I definitely think it's a good idea to write automated tests for these changes since they're pretty complex and super easy to break

**Comment by Beamanator** [comment]
Sounds like a great idea to me! 👍 

I thinkkk I'll create a new issue for this tomorrow, and I'll assign you @nkdengineer 🚀

**Comment by Beamanator** [comment]
Ummmm pretty sure @puneetlath is BZ here? 😅

**Comment by laurenreidexpensify** [comment]
@puneetlath isn't on the team anymore :)

**Comment by Beamanator** [comment]
ohhhhhh

**Comment by laurenreidexpensify** [comment]
That said, I'm locked out of my Upwork account right now so @puneetlath can you take this one to the finish line 🙏 thanks!

**Comment by puneetlath** [comment]
Payment summary: 
- C - $500 - @nkdengineer - paid via Upwork
- C+ - $500 - @ntdiary - to be paid via NewDot

Thanks everyone!

@Beamanator I'll leave the issue open for you to close once you've created the follow-up issue.

**Comment by Beamanator** [comment]
Created follow-up: https://github.com/Expensify/App/issues/55003

@nkdengineer can you please comment there so i can assign you? Closing this one now 👍
